Quaternion/Quaternion.py

```diff
@@ -0,0 +1,513 @@
+
+       >>> q2 = q1 * dq
+       >>> dq = q1.inv() * q2
+
+     Any permutation of ``q1 / q2`` using the divide operator is WRONG.  One can also
+     use the ``dq`` method::
+
+       >>> dq = q1.dq(q2)
+
+    :param attitude: initialization attitude for quat
+
+    ``attitude`` may be:
+      * another Quat
+      * a 4 element array (expects x,y,z,w quat form)
+      * a 3 element array (expects ra,dec,roll in degrees)
+      * a 3x3 transform/rotation matrix
+
+    """
+
+    def __init__(self, attitude):
+        self._q = None
+        self._equatorial = None
+        self._T = None
+        # checks to see if we've been passed a Quat
+        if isinstance(attitude, Quat):
+            self._set_q(attitude.q)
+        else:
+            # make it an array and check to see if it is a supported shape
+            attitude = np.array(attitude)
+            if len(attitude) == 4:
+                self._set_q(attitude)
+            elif attitude.shape == (3, 3):
+                self._set_transform(attitude)
+            elif attitude.shape == (3,):
+                self._set_equatorial(attitude)
+            else:
+                raise TypeError(
+                    "attitude is not one of possible types (3 or 4 elements, Quat, or 3x3 matrix)")
+
+    def _set_q(self, q):
+        """
+        Set the value of the 4 element quaternion vector
+
+        :param q: list or array of normalized quaternion elements
+        """
+        q = np.array(q)
+        if abs(np.sum(q**2) - 1.0) > 1e-6:
+            raise ValueError(
+                'Quaternion must be normalized so sum(q**2) == 1; use Quaternion.normalize')
+        self._q = (q if q[3] > 0 else -q)
+        # Erase internal values of other representations
+        self._equatorial = None
+        self._T = None
+
+    def _get_q(self):
+        """
+        Retrieve 4-vector of quaternion elements in [x, y, z, w] form
+
+        :rtype: numpy array
+
+        """
+        if self._q is None:
+            # Figure out q from available values, doing nothing others are not defined
+            if self._equatorial is not None:
+                self._q = self._equatorial2quat()
+            elif self._T is not None:
+                self._q = self._transform2quat()
+        return self._q
+
+    # use property to make this get/set automatic
+    q = property(_get_q, _set_q)
+
+    def __repr__(self):
+        q = self.q
+        return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
+                .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+
+    def _set_equatorial(self, equatorial):
+        """Set the value of the 3 element equatorial coordinate list [RA,Dec,Roll]
+           expects values in degrees
+           bounds are not checked
+
+           :param equatorial: list or array [ RA, Dec, Roll] in degrees
+
+        """
+        att = np.array(equatorial)
+        ra, dec, roll = att
+        self._ra0 = ra
+        if (ra > 180):
```

Comment:
I think if we added

```
self._roll0 = roll
```

here when the equatorial representation is initialized, we'd never need "get_zero" (though the next few lines could be replaced with get_zero calls).